### ISSUE TITLE
Make Dropout take a dim=... argument

### ISSUE THREAD

**Issue by thomasahle** [pytorch/pytorch#109582]
### 🚀 The feature, motivation and pitch

Currently PyTorch has Dropout, Dropout1d, Dropout2d and Dropout3d, for handling data with shape either shape (*, D), (N, C, D), (N, C, W, H) or (N, C, W, H, Z).
However none of them support the case where I want to do channel dropout but have shape, say, (B, N, C, D).
It would be much easier if the standard Dropout class just took a `dim` argument for specifying the level at which to do dropout.
For example, `dim=-1` could be the default, and would correspond to the current Dropout behavior.
`dim=-2` would be like Dropout1d, but support more than one batch dimensions, e.g. (B, N, C, D).

**Comment by mikaylagawarecki** [comment]
Hey, thanks for the issue! Does the following give the desired behavior here?

```python
import torch

dropout = torch.nn.Dropout1d()
# assume input has shape (B, N, C, D)
input = input.view(-1, C, D)
dropout(input)
input.view(B, N, C, D)
```

**Comment by thomasahle** [comment]
Yeah, I'm doing this now:
```
class ChannelDropout(nn.Module):
    def __init__(self, p=0.5):
        super(ChannelDropout, self).__init__()
        self.p = p

    def forward(self, x):
        x_reshaped = x.reshape((-1,) + x.shape[-2:])
        x_dropped = F.dropout1d(x_reshaped, self.p, self.training, inplace=False)
        return x_dropped.reshape(*x.shape)
```
But it doesn't work if the channel dimension is at location other than -2.
It would be nice if I could just specify the channel dimension as an argument to Dropout.

**Comment by mikaylagawarecki** [comment]
Hm can you specify it to your `ChannelDropout` class as

```python

class ChannelDropout(nn.Module):
    def __init__(self, p=0.5, dim=-1):
        super(ChannelDropout, self).__init__()
        self.p = p

    def forward(self, x):
        x_reshaped = x.reshape((-1,) + x.shape[dim:])
        x_dropped = F.dropout1d(x_reshaped, self.p, self.training, inplace=False)
        return x_dropped.reshape(*x.shape)

```

**Comment by thomasahle** [comment]
No, because dropout1d only accepts input sizes (N, C, L) or (C, L).
But I think something like
```
x_reshaped = x.flatten(0, dim-1).flatten(dim+1, -1)
```
would work.

Such a class would also be able to replace Dropout2d and Dropout3d.
